cobalt/browser/h5vcc_experiments/h5vcc_experiments_impl.cc

```diff
@@ -0,0 +1,156 @@
+        // TODO(b/407734134): Register UMA for non-parsable features.
+        LOG(ERROR) << "Unable to parse disabled feature: " << disabled;
+        continue;
+      }
+      override_map[feature] =
+          h5vcc_experiments::mojom::OverrideState::OVERRIDE_DISABLE_FEATURE;
+    }
+  }
+  auto it = override_map.find(feature_name);
+  if (it != override_map.end()) {
+    return it->second;
+  }
+  return h5vcc_experiments::mojom::OverrideState::OVERRIDE_USE_DEFAULT;
+}
+
+std::string GetFeatureParamInternal(const std::string& feature_param_name) {
+  return base::GetFieldTrialParamValue(cobalt::kCobaltExperimentName,
+                                       feature_param_name);
+}
+
+}  // namespace
+
+// TODO (b/395126160): refactor mojom implementation on Android
+H5vccExperimentsImpl::H5vccExperimentsImpl(
+    content::RenderFrameHost& render_frame_host,
+    mojo::PendingReceiver<mojom::H5vccExperiments> receiver)
+    : content::DocumentService<mojom::H5vccExperiments>(render_frame_host,
+                                                        std::move(receiver)) {}
+
+void H5vccExperimentsImpl::Create(
+    content::RenderFrameHost* render_frame_host,
+    mojo::PendingReceiver<mojom::H5vccExperiments> receiver) {
+  new H5vccExperimentsImpl(*render_frame_host, std::move(receiver));
+}
+
+void H5vccExperimentsImpl::SetExperimentState(
+    base::Value::Dict experiment_config,
+    SetExperimentStateCallback callback) {
+  auto experiment_config_ptr =
+      cobalt::GlobalFeatures::GetInstance()->experiment_config();
+
+  experiment_config_ptr->SetDict(
+      cobalt::kExperimentConfigFeatures,
+      std::move(experiment_config.Find(cobalt::kExperimentConfigFeatures)
+                    ->GetDict()));
+  experiment_config_ptr->SetDict(
+      cobalt::kExperimentConfigFeatureParams,
+      std::move(experiment_config.Find(cobalt::kExperimentConfigFeatureParams)
+                    ->GetDict()));
+  experiment_config_ptr->SetList(
+      cobalt::kExperimentConfigExpIds,
+      std::move(
+          experiment_config.Find(cobalt::kExperimentConfigExpIds)->GetList()));
+  std::move(callback).Run();
+}
+
+void H5vccExperimentsImpl::GetFeature(const std::string& feature_name,
+                                      GetFeatureCallback callback) {
+  std::move(callback).Run(GetFeatureInternal(override_map_, feature_name));
+}
+
+void H5vccExperimentsImpl::GetFeatureParam(
+    const std::string& feature_param_name,
+    GetFeatureParamCallback callback) {
+  std::move(callback).Run(GetFeatureParamInternal(feature_param_name));
+}
+
+void H5vccExperimentsImpl::ResetExperimentState(
+    ResetExperimentStateCallback callback) {
+  PrefService* experiment_config =
+      cobalt::GlobalFeatures::GetInstance()->experiment_config();
+  experiment_config->ClearPref(cobalt::kExperimentConfigFeatures);
+  experiment_config->ClearPref(cobalt::kExperimentConfigFeatureParams);
+  experiment_config->ClearPref(cobalt::kExperimentConfigExpIds);
```

Comment:
Doesn't PrefService have a mechanism to force a JSON write to disk here? Like a flush? I think we should do that here so that state is wiped immediately. I believe, by default, PrefService asynchronously writes to disk and I wouldn't want to risk missing this reset as it can be a very important operation.

Reply:
Are you referring to CommitPendingWrite()?